### PR TITLE
unified group expression refs and expression refs, removed unnecessar…

### PR DESCRIPTION
**Pull request by normen662**
…y type parameter

### CODE REVIEW

**Review by hatyo — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java [hunk 1]
```diff
@@ -52,28 +53,13 @@
...
- *
- * <p>
- * The reference abstraction is designed to make it difficult for authors of rules to mutate group expressions directly,
- * which is undefined behavior. Note that a {@link GroupExpressionRef} cannot be "dereferenced" using the {@link #get()}
- * method if it contains more than one member. Expressions with more than one member should not be used outside of the
- * query planner, and {@link #get()} should not be used inside the query planner.
- * </p>
- * @param <T> the type of planner expression that is contained in this reference
+ * This interface is used mostly as an (admittedly surmountable) barrier to rules mutating bound references directly,
+ * which is undefined behavior.
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java:57]: done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java [hunk 2]
```diff
@@ -1,9 +1,9 @@
 /*
- * GroupExpressionRef.java
+ * Reference.java
  *
  * This source file is part of the FoundationDB open source project
  *
- * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ * Copyright 2015-2018 Apple Inc. and the FoundationDB project authors
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java:6]: done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java [hunk 3]
```diff
@@ -96,17 +82,31 @@ private GroupExpressionRef(@Nonnull LinkedIdentitySet<T> members) {
...
 
+    /**
+     * Return the {@link RecordQueryPlan} contained in this reference.
+     * @return the {@link RecordQueryPlan} contained in this reference
+     * @throws UngettableReferenceException if the reference does not support retrieving its expression
+     */
     @Nonnull
-    @Override
-    public T get() {
+    public RecordQueryPlan getAsPlan() {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java:91]: done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java [hunk 4]
```diff
@@ -580,27 +585,42 @@ private static boolean containsInMember(@Nonnull final RelationalExpression memb
...
+        LinkedIdentitySet<RelationalExpression> members = new LinkedIdentitySet<>();
         expressions.forEach(Debugger::registerExpression);
         members.addAll(expressions);
-        return new GroupExpressionRef<>(members);
+        return new Reference(members);
+    }
+
+    /**
+     * An exception thrown when {@link #get()} is called on a reference that does not support it, such as a group reference.
+     * A client that encounters this exception is buggy and should not try to recover any information from the reference
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java:613]: This means to say, it throws if there is more than one in the group. I changed that to be clear.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java [hunk 5]
```diff
@@ -580,27 +585,42 @@ private static boolean containsInMember(@Nonnull final RelationalExpression memb
...
+     * that threw it.
+     * <br>
+     * This exceeds the normal maximum inheritance depth because of the depth of the <code>RecordCoreException</code>
+     * hierarchy.
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Reference.java:617]: Yeah that seems to be nonsense.

**Review by hatyo — DISMISSED** [review]
LGTM

**Review by hatyo — APPROVED** [review]

**Pull request merged**